### DISCUSSION TITLE
Patrol for Flutter web apps

### DISCUSSION THREAD

**Discussion by pandoras-toolbox** [leancodepl/patrol#1489] [Q&A] [answered]
I think I read it, can't find where, that Patrol does not support Flutter web apps but is going to.

It would be useful for us because we also have Flutter web apps needing test automation.

When will Patrol have support for testing Flutter web apps? Like in a few months or so?

**Comment by NupurSwS** [comment]
Following the thread...

**Comment by bartekpacia** [comment]
Hi, currently we don't have any plans to add support for Flutter Web. If anything, we'll probably ship Flutter Desktop earlier.

> Like in a few months or so?

At least.

> Flutter web apps needing test automation.

What sort of automation? The same as on mobile - native permissions etc?

**Reply by jay-athelas** [reply]
what do you mean what sort of automation, the same automation any web app needs, tf

**Comment by pandoras-toolbox** [comment]
> What sort of automation? The same as on mobile - native permissions 

I asked because in [How We 10x Improved Flutter UI Testing With Patrol 2.0](https://leancode.co/blog/patrol-2-0-improved-flutter-ui-testing) a lot of problems are mentioned regarding the integration_test plugin, especially in section "integration_test doesn’t deliver".

My hope was that we would not have these problems if we could use Patrol for testing Flutter web apps.  It was not because of any native features, which we do not need for web app testing I think.

**Reply by bartekpacia** [reply, marked as answer]
Ah, okay - thanks for explaining. Honestly, I have never used `integration_test` when testing a Flutter Web app, so I don't really know if those problems are present in that scenario.

Web is very different from mobile and I have no experience with it. I'd have to investigate how things look like there.

> It was not because of any native features, which we do not need for web app testing I think.

Yeah, I think the same. That's what always baffles me when people ask about Patrol on Flutter Web - like, "c'mon, what's the use case?" :)

**Reply by pandoras-toolbox** [reply]
I cannot confirm these problems yet for Flutter web apps which are mentioned in the "How we 10x ..." blog post.

I just tried a little proof of concept to see if Flutter web apps can be tested with the integration_test package, which worked. That is my only experience with it so far.

It does not work to automate Flutter web apps with Selenium. But I haven't given up, because that would have a lot of advantages. An option could be to make use of the Dart/JavaScript interoperability and to use the Flutter locators and actions through JavaScript by Selenium.

For us it is still not decided how we will automate Flutter web app tests, what is the best way for us to do it, or even a good way.

**Reply by jagadeesht9** [reply]
Any update on this ?? As I am also facing this issue and need help in finding an automation tool which could automate Android, IOS and web Flutter app. @pandoras-toolbox

**Reply by desmond206x** [reply]
+1

**Reply by ecolab-neil** [reply]
@jagadeesht9 @pandoras-toolbox @desmond206x @bartekpacia  any update or thoughts re: flutter web with patrol or even maestro?
is there anything that would "need" "native auatomatioN" - wwhat would ther eexperience with flutter_integraiton and patrol-finders only be like?

**Comment by pandoras-toolbox** [comment]
@jagadeesht9 In my opinion there is no good enough test automation tool for Flutter web apps, sad but true. In our projects I suggest to automate tests on API level and to manually tests Flutter web apps. For mobile apps however I think Patrol is a good test automation tool. In my opinion Flutter web apps should not even be developed because there is no good enough E2E test tool for it.

**Comment by RashmiShashindra** [comment]
Hi @bartekpacia ,

I have a use case of running the same flutter tests on both Mobile and Web. Because our app is being built on Android Native, Android Web, iOS native, iOS Web and Some customer base use it via Desktop Web too. Currently we have the automation for the Android and iOS Mobile Automation using Patrol But we are unable to test the Web version via automation. Could you please look into this as a priority as per my thinking Many of Patrol users face the same problem. If you are moving forward with this Web Support, may I know if there is any ETA for the feature release?

Thanks and Best!

**Reply by ecolab-neil** [reply]
@RashmiShashindra  any update in general -what did you end up doing for flutter web?

**Comment by JaredEzz** [comment]
I have a use case for Flutter Web as well. At ostrichcyber-risk.com, we're currently using `integration_test` and it falls short in so many ways, causing us to have to do manual testing and substitute with more unit testing and API testing than if we had a good intergration testing framework for web. 

As for the use case, it doesn't differ too much from mobile apps. Tap on things, automate inputs, allow expectations of outputs, and do it in a repeatable way that can be automated as part of CI/CD build pipelines with good reporting.

If Patrol could do that, we'd be instant customers.

**Reply by ecolab-neil** [reply]
@JaredEzz  any updates onthis? it's nice that you have more details, i'm curious about what ways fall short - what would patrol_finders not do?

**Comment by zoskar** [comment]
Web support has been introduced in Patrol 4.0
https://pub.dev/packages/patrol/changelog#400

**Reply by giorgio79** [reply]
So does this support Flutter web that uses the canvas?

**Reply by piotruela** [reply]
Do you mean canvaskit as a renderer? If so, yes it works

**Reply by giorgio79** [reply]
Fantastic news thx
